docs/settings.md
# Settings

Here are all available settings with their defaults, you can override them in your project settings
```python
DRF_ERROR_HANDLER = {
    # class responsible for handling the exceptions. Can be subclassed to change
    # which exceptions are handled by default, to update which exceptions are
    # reported to error monitoring tools (like Sentry), ...
    "EXCEPTION_HANDLER_CLASS": "drf_error_handler.handler.ExceptionHandler",
    # class responsible for generating error response output. Can be subclassed
    # to change the format of the error response.
    "EXCEPTION_FORMATTER_CLASS": "drf_error_handler.formatter.ExceptionFormatter",
    # When a validation error is raised in a nested serializer, the 'attr' key
    # of the error response will look like:
    # {field}{NESTED_FIELD_SEPARATOR}{nested_field}
    # for example: 'shipping_address.zipcode'
    "NESTED_FIELD_SEPARATOR": ".",

    # The below settings are for OpenAPI 3 schema generation

    # ONLY the responses that correspond to these status codes will appear
    # in the API schema.
    "ALLOWED_ERROR_STATUS_CODES": [
        "400",
        "401",
        "403",
        "404",
        "405",
        "406",
        "415",
        "429",
        "500",
    ],

    # A mapping used to override the default serializers used to describe
    # the error response. The key is the status code and the value is a string
    # that represents the path to the serializer class that describes the
    # error response.
    "ERROR_SCHEMAS": None,

    # When there is a validation error in list serializers, the "attr" returned
    # will be sth like "0.email", "1.email", "2.email", ... So, to describe
    # the error codes linked to the same field in a list serializer, the field
    # will appear in the schema with the name "INDEX.email"
    "LIST_INDEX_IN_API_SCHEMA": "INDEX",

    # When there is a validation error in a DictField with the name "extra_data",
    # the "attr" returned will be sth like "extra_data.<key1>", "extra_data.<key2>",
    # "extra_data.<key3>", ... Since the keys of a DictField are not predetermined,
    # this setting is used as a common name to be used in the API schema. So, the
    # corresponding "attr" value for the previous example will be "extra_data.KEY"
    "DICT_KEY_IN_API_SCHEMA": "KEY",

    # should be unique to error components since it is used to identify error
    # components generated dynamically to exclude them from being processed by
    # the postprocessing hook. This avoids raising warnings for "code" and "attr"
    # which can have the same choices across multiple serializers.
    "ERROR_COMPONENT_NAME_SUFFIX": "ErrorComponent",
}
```


docs/changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [UNRELEASED]
### Added
- declare support for DRF 3.15

### Fixed
- enforce support of only drf-spectacular 0.27 and newer in pyproject.toml
- ensure examples from `@extend_schema_serializer` are not ignored when adding error response examples

## [0.13.0] - 2024-02-28
### Changed
- If you're using drf-spectacular 0.27.0 or newer, update `ENUM_NAME_OVERRIDES` entries to reference `choices`
rather than `values`. The list of overrides specific to this package should become like this:
```python
SPECTACULAR_SETTINGS = {
    # other settings
    "ENUM_NAME_OVERRIDES": {
        "ValidationErrorEnum": "drf_error_handler.openapi_serializers.ValidationErrorEnum.choices",
        "ClientErrorEnum": "drf_error_handler.openapi_serializers.ClientErrorEnum.choices",
        "ServerErrorEnum": "drf_error_handler.openapi_serializers.ServerErrorEnum.choices",
        "ErrorCode401Enum": "drf_error_handler.openapi_serializers.ErrorCode401Enum.choices",
        "ErrorCode403Enum": "drf_error_handler.openapi_serializers.ErrorCode403Enum.choices",
        "ErrorCode404Enum": "drf_error_handler.openapi_serializers.ErrorCode404Enum.choices",
        "ErrorCode405Enum": "drf_error_handler.openapi_serializers.ErrorCode405Enum.choices",
        "ErrorCode406Enum": "drf_error_handler.openapi_serializers.ErrorCode406Enum.choices",
        "ErrorCode415Enum": "drf_error_handler.openapi_serializers.ErrorCode415Enum.choices",
        "ErrorCode429Enum": "drf_error_handler.openapi_serializers.ErrorCode429Enum.choices",
        "ErrorCode500Enum": "drf_error_handler.openapi_serializers.ErrorCode500Enum.choices",
        # other overrides
    },
}
```

### Added
- add compatibility with drf-spectacular 0.27.x
- add support for django 5.0

### Fixed
- Ensure accurate traceback inclusion in 500 error emails sent to ADMINS by capturing the original exception information using `self.exc`. This fixes the issue where tracebacks were previously showing as None for `django version >= 4.1`.
- Handle error responses with +1000 errors

## [0.12.6] - 2023-10-25
### Added
- declare support for type checking
- add support for django 4.2
- add support for python 3.12

### Fixed
- Avoid calling `AutoSchema.get_request_serializer` when inspecting a get operation for possible error responses.

## [0.12.5] - 2023-01-14
### Added
- allow adding extra validation errors on an operation-basis using the new `@extend_validation_errors` decorator.
You can find [more information about that in the documentation](openapi.md#customize-error-codes-on-an-operation-basis).

### Fixed
- use `model._default_manager` instead of `model.objects`.
- Don't generate error responses for OpenAPI callbacks.
- Make `_should_add_http403_error_response` check if permission is `IsAuthenticated` and 
`AllowAny` via `type` instead of `isinstance`
- Don't collect error codes from nested `read_only` fields

## [0.12.4] - 2022-12-11
### Fixed
- account for specifying the request serializer as a basic type (like `OpenApiTypes.STR`) or as a
`PolymorphicProxySerializer` using `@extend_schema(request=...)` when determining error codes for validation errors.

## [0.12.3] - 2022-11-13
### Added
- add support for python 3.11

## [0.12.2] - 2022-09-25
### Added
- When a custom validator class defines a `code` attribute, add it to the list of error codes of raised by
the corresponding field.
- add support for DRF 3.14

## [0.12.1] - 2022-09-03
### Fixed
- generate the mapping for discriminator fields properly instead of showing a "null" value in the generated schema (#12).

## [0.12.0] - 2022-08-27
### Added
- add support for automatically generating error responses schema with [drf-spectacular](https://github.com/tfranzel/drf-spectacular).
Check out the [corresponding documentation page](https://drf-error-handler.readthedocs.io/en/latest/openapi.html)
to know more about the integration with drf-spectacular.
- add support for django 4.1

## [0.11.0] - 2022-06-24
### Changed (Backward-incompatible)
- Removed all imports from `drf_error_handler.__init__.py`. This avoids facing the `AppRegistryNotReady` error
in certain situations (fixes #7). This change **only affects where functions/classes are imported from**, there are
**no changes to how they work**. To upgrade to this version, you need to:
  - Update the `"EXCEPTION_HANDLER"` setting in `REST_FRAMEWORK` to `"drf_error_handler.handler.exception_handler"`.
  - If you imported the exception handler directly, make sure the import looks like this
  `from drf_error_handler.handler import exception_handler`.
  - If you imported the exception handler class, make sure the import looks like this
  `from drf_error_handler.handler import ExceptionHandler`.
  - If you imported the exception formatter class, make sure the import looks like this
  `from drf_error_handler.formatter import ExceptionFormatter`.

## [0.10.2] - 2022-05-08
### Fixed
- disable tag creation by the "create GitHub release" action since it is already created by tbump

## [0.10.1] - 2022-05-08
### Fixed
- add write permission to create release action, so it can push release notes to GitHub
- fix license badge link so it works on PyPI

## [0.10.0] - 2022-05-08
### Added

- Build the documentation automatically on every commit to the main branch. The docs are
[hosted on readthedocs](https://drf-error-handler.readthedocs.io/en/latest/).
- Add package metadata
- add a GitHub workflow to create a GitHub release when a new tag is pushed
- add a GitHub workflow to run tests on every push and pull request
- add test coverage

## [0.9.0] - 2022-05-07
### Added

- Common error response format for DRF-based APIs
- Easily customize the error response format.
- Handle error responses for list serializers and nested serializers. 
- Add documentation
- Add tests
- Automate release steps


drf_standardized_errors/settings.py
from typing import Any, Dict, Optional, Set, Tuple

from django.conf import settings
from django.core.signals import setting_changed
from django.dispatch import receiver
from rest_framework.settings import import_from_string, perform_import


class PackageSettings:
    """
    Copy of DRF APISettings class with support for importing settings that
    are dicts with value as a string representing the path to the class
    to be imported.
    """

    setting_name = "DRF_STANDARDIZED_ERRORS"

    def __init__(
        self,
        defaults: Optional[Dict[str, Any]] = None,
        import_strings: Optional[Tuple[str, ...]] = None,
    ):
        self.defaults = defaults or DEFAULTS
        self.import_strings = import_strings or IMPORT_STRINGS
        self._cached_attrs: Set[str] = set()

    @property
    def user_settings(self) -> Dict[str, Any]:
        if not hasattr(self, "_user_settings"):
            self._user_settings = getattr(settings, self.setting_name, {})
        return self._user_settings

    def __getattr__(self, attr: str) -> Any:
        if attr not in self.defaults:
            raise AttributeError("Invalid API setting: '%s'" % attr)

        try:
            # Check if present in user settings
            val = self.user_settings[attr]
        except KeyError:
            # Fall back to defaults
            val = self.defaults[attr]

        # Coerce import strings into classes
        if attr in self.import_strings:
            if isinstance(val, dict):
                val = {
                    status_code: import_from_string(error_schema, attr)
                    for status_code, error_schema in val.items()
                }
            else:
                val = perform_import(val, attr)

        # Cache the result
        self._cached_attrs.add(attr)
        setattr(self, attr, val)
        return val

    def reload(self) -> None:
        for attr in self._cached_attrs:
            delattr(self, attr)
        self._cached_attrs.clear()
        if hasattr(self, "_user_settings"):
            delattr(self, "_user_settings")


DEFAULTS: Dict[str, Any] = {
    "EXCEPTION_HANDLER_CLASS": "drf_standardized_errors.handler.ExceptionHandler",
    "EXCEPTION_FORMATTER_CLASS": "drf_standardized_errors.formatter.ExceptionFormatter",
    "ENABLE_IN_DEBUG_FOR_UNHANDLED_EXCEPTIONS": False,
    "NESTED_FIELD_SEPARATOR": ".",
    "ALLOWED_ERROR_STATUS_CODES": [
        "400",
        "401",
        "403",
        "404",
        "405",
        "406",
        "415",
        "429",
        "500",
    ],
    "ERROR_SCHEMAS": None,
    "LIST_INDEX_IN_API_SCHEMA": "INDEX",
    "DICT_KEY_IN_API_SCHEMA": "KEY",
    "ERROR_COMPONENT_NAME_SUFFIX": "ErrorComponent",
    
    "EXCEPTION_RESPONSE_BUSINESS_ATTRIBUTE": "b_code",
    
    "VALIDATION_ERROR_BUSINESS_STATUS_CODE": 1001,
    "PARSE_ERROR_BUSINESS_STATUS_CODE": 1002,
    "AUTHENTICATION_FAILED_BUSINESS_STATUS_CODE": 1003,
    "NOT_AUTHENTICATION_BUSINESS_STATUS_CODE": 1004,
    "PERMISSION_DENIED_BUSINESS_STATUS_CODE": 1005,
    "NOT_FOUND_BUSINESS_STATUS_CODE": 1006,
    "METHOD_NOT_ALLOWED_BUSINESS_STATUS_CODE": 1007,
    "NOT_ACCEPTABLE_BUSINESS_STATUS_CODE": 1008,
    "UNSUPPORTED_MEDIA_TYPE_BUSINESS_STATUS_CODE": 1009,
    "THROTTLED_BUSINESS_STATUS_CODE": 1010,
}

IMPORT_STRINGS = (
    "EXCEPTION_FORMATTER_CLASS",
    "EXCEPTION_HANDLER_CLASS",
    "ERROR_SCHEMAS",
)

package_settings = PackageSettings(DEFAULTS, IMPORT_STRINGS)

@receiver(setting_changed)
def reload_package_settings(*args: Any, **kwargs: Any) -> None:
    setting = kwargs["setting"]
    if setting == package_settings.setting_name:
        package_settings.reload()
